src/geometry/types.rs

```diff
@@ -121,6 +122,89 @@ impl Vector {
     }
 }
 
+// Rectangle /////////////////////////////////////////////////////////////////
+#[derive(PartialEq, Debug)]
+pub struct Size {
+    pub width: f32,
+    pub height: f32,
+}
+
+#[derive(PartialEq, Debug)]
+pub struct Rect {
+    // the leftmost coordination system of the rectangle
+    pub origin: Point,
+    // the width and high of the rectangle
+    pub size: Size,
+}
+
+impl Rect {
+
+    pub fn half_size(&mut self) {
+        self.size.width = self.size.width/2.0;
+        self.size.height = self.size.height/2.0;
+    }
+
+    fn quarter_sized(&self) -> Rect{
+        Rect{
+            origin: Point{
+                x: self.origin.x,
+                y: self.origin.y,
+            },
+            size: Size{
+                width: self.size.width/ 2.0,
+                height: self.size.height/2.0,
+            }
+        }
+    }
+
+    pub fn quadrants(&self) -> (Rect, Rect, Rect, Rect) {
+
+        //the upper left rectangle
+        let mut rect1 = self.quarter_sized();
+
+        //the upper right rectangle
+        let mut rect2 = self.quarter_sized();
+
+        rect2.origin.x = rect2.origin.x + rect2.size.width;
+
+        //the lower right rectangle
+        let mut rect3 = self.quarter_sized();
+        rect3.origin.x = rect3.origin.x + rect3.size.width;
+        rect3.origin.y = rect3.origin.y + rect3.size.height;
+
+        //the lower left rectangle
+        let mut rect4 = self.quarter_sized();
+        rect4.origin.y = rect4.origin.y + rect4.size.height;
+
+        let rects: (Rect, Rect, Rect, Rect) = (rect1, rect2, rect3, rect4);
```

Comment:
You don't need to specify the type of the type here, the compiler can infer it. Also for clarity, I would suggest to give better names than `rect1` and `rect2`, something like `upper_left` and `upper_right`.